### ISSUE TITLE
Add function to reset the parameters of the network

### ISSUE THREAD

**Issue by hlin117** [Lasagne/Lasagne#637]
Apologies if there already is a mechanism for resetting the network; nothing seemed to come up on google for me.

With all of the functions to [set and get](http://lasagne.readthedocs.org/en/latest/modules/layers.html#module-lasagne.layers) values from the network, it would be nice to have a way to reset the values as well. 

This function would be great for `k`-fold cross validation; it'll provide a clean interface for resetting the parameters right before training another fold in the evaluation process.

**Comment by benanne** [comment]
This might be a bit tricky as it requires storing the initializers for future use, which is currently not done. So it's not a small change, probably. Is there a reason why you can't simply create a function that instantiates a new copy of the network? Just trying to get an idea of the use cases for this.

**Comment by hlin117** [comment]
Right now, that's my current approach. (In fact I created a class for my
network.) However, I only redesigned my code because there was no
reset_params function. So, work arounds are definitely possible, but it's
not preferable.

I wasn't aware that it wasn't a quick change.

**Comment by benanne** [comment]
I'm actually mistaken -- most of the logic should be in `Layer.add_param`, so we could modify that function to store the initializers somewhere. So it is a more local change than I thought at first.

However, Lasagne allows different ways to specify initializers, and we would need to clearly define what it means to "reset" all of them:
* initialization function: this one is easy, just call the function again to generate a new set of values
* explicit initial value (numpy array): just reuse the initial value
* Theano shared variable or expression: ???

I don't really know what the correct way is to handle that last one. I think there is basically no clean solution for that. We can't really change the contents of the shared variables as they may also be used elsewhere that we don't know of.

So that means `reset_params()` would not work for any nets with parameter sharing, for example... which is a bit icky. So I think I'm still not in favour of adding it. Wrapping the net in a class or function is probably the best solution.

**Comment by f0k** [comment]
Thanks for your suggestion!

I'm also in favour for leaving that up to the user, as there's no clear general solution for the third case, and the second case will raise memory usage unneededly for users that don't require resettable weights (including storage needs for pickled network models).
Rebuilding the network or calling `get_all_param_values()` right after initialization (if resetting to *exactly* the same parameters is fine) should be easy enough!

Closing this issue (feel free to continue the discussion though).

**Comment by hlin117** [comment]
Thanks everyone for this helpful discussion!

**Comment by Erotemic** [comment]
Joining this conversation from #740. I'm in favor of adding a feature that allows a layer optionally  remember how its parameters were initialized. I think adding reset_params is a good use case for this feature. However, I think reset_params should fully work only if the network was defined with a  lasagne.init.Initializer. 

Considering the 3 cases for initial values: 
> 1) initialization function
> 2) explicit initial value (numpy array)
> 3) Theano shared variable or expression
> 

Let me present an argument for why this persistent initializers might be worth implementing only to handle the first case. 

Even if layers have the ability to remember there initialization method, I think it is reasonable to not require that they have to remember it. In this way layers can remember the simple implicit (function based) initialization values and can leave the more difficult explicit (numpy array / shared variable) to the user. 

**Why the first case is worth supporting:**
I think the first case is by far the most common. It is also fairly extensible because a user can subclass lasange.init.Initializer. The memory footprint is only as big as the Initializer itself, which for all standard subclasses is extremely small. Of course custom initializers can implement arbitrary code, which could make a very memory intensive Initializer, but the user is aware that at some point this initializer will be in memory and be used. I don't think much of a problem is added if this hypothetical Initializer persists, and if it is a problem, it should be possible for a layer to forget its initialization state.

**Why the second and third case should not be supported:**
The second and third case are equivalent to loading pretrained weights for specific layers.
When dealing with pretrained weights, the user must always explicitly load them and set their value.
It should be no different when using weights that have an explicitly set state that can not be expressed as a simple function. 

Of course, if the user desperately wanted their custom weights to be remembered they could subclass an Initializer object and the layer would remember it. However, this would incur the memory penalty stated above. I think this is ok, because the user is aware of creating a class with this large memory footprint and letting it persist in a layer.

**Pros/Cons of implementing a solution that only addresses the first case:**

* (Negative) nonstandard callable / explicit initialization wont work.

The main objection I see is that a reset_params method would not work on networks with parameter sharing. I think this can easily be handled by simply raising an exception if any layer is missing an initializer. An error message explaining that reset_params only remembers standard initialization methods could be raised, and I don't think that would cause too many headaches. Perhaps a flag could even be set to ignore those layers and reset params for only layers with "resetable" parameters ("resetable" might even be a tag that can be applied to params to denote if it knows about standard Initializer).  

* (Positive) The vast majority of networks will work with reset_params.

Most networks will use params initialized from an object that inherits Initializer because they are the default. Even when the a param is not defaulted, it is most often changed to another subclass of Initializer (like HeNormal or Orthogonal).

* (Positive) The network architecture can be fully written to disk.

In the case that the code that defines a network definition is lost, then there is no way to easily recreate the network. Having the initializer saved (if it is standard) allows for a portable architecture file to be written and live outside python code. 

* (Positive) Case 2 and 3 can be supported if the user wraps it with an Initializer object. 

Quoting from the Initializer documentation 
>     The :class:`Initializer` class represents a weight initializer used
>     to initialize weight parameters in a neural network layer. It should be
>     subclassed when implementing new types of weight initializers.

Having a quick and dirty way to support setting explicit values is nice, but in most cases is it preferable to implement a custom Initializer. The reset_params function should only support these standard classes. When subclassing Initializer it because the users job to ensure that it is memory efficient (maybe by keeping the weights on disk and only loading them when sample is called). 

* (Positive) Lazy initialization is a natural extension.

Another consequence of implementing this change is that layer parameters don't need to be initialized immediately if the layer knows how to do it. This removes the overhead of creating each parameter when you are simply defining your network. In many cases you I'm simply testing code that requires my layers to be defined, but I have no intention of ever executing the network. However, in these cases I still have to wait while SVD executes if I use orthogonal initialization. I believe all layer parameters currently default to some standard initialization function like GlorotUniform. Most layers in most networks will be initialized using these standard functions. In the cases that they are not the user can lazily load custom parameters into those layers as they would with pretrained weights.

**Comment by f0k** [comment]
Thank you for the elaborate discussion!

> The main objection I see is that a reset_params method would not work on networks with parameter sharing.

Depends on the implementation. If we limit support to initializers returning a numpy array, we can just call `set_value()` on the original shared variable (instead of creating a new one) and all parameter sharing is kept intact. So that negative point can be addressed.

My main objection is that adding a resetting feature that only resets *some* of the weights may lead users into running wrong experiments. But maybe users with custom initializers will become suspicious at how this could possibly work and have a closer look at the documentation.

> "resetable" might even be a tag that can be applied to params to denote if it knows about standard Initializer

Sounds good to me (but I'd prefer `resettable`).

> When dealing with pretrained weights, the user must always explicitly load them and set their value.

The user could call `get_all_param_values` and `set_all_param_values` with `resettable=False` to remember and reset such values if needed.

> Another consequence of implementing this change is that layer parameters don't need to be initialized immediately if the layer knows how to do it.

If implemented, this should be optional. Some existing user code may rely on the weights being initialized exactly when constructing the layer, and changing things will make previous results irreproducible (think of a seeded RNG that will give different results if anything happens in a different order).
In any case, the implementation could use `init.Constant()` for all resettable weights so we create the necessary shared variables, and then rely on `reset_params` to initialize the weights when needed. I think there's no other way to implement this without touching existing layers -- they expect `add_param()` to return a shared variable that is valid for the layer's lifetime.


Now for a name discussion: `reset_params` sounds like it could restore the original state of the network. Maybe `reinitialize_params` would be more suitable. A `reinitializable` tag sounds awkward, though. Is there any third solution?

For the implementation, I'll comment in #740.

**Comment by Erotemic** [comment]
> Maybe reinitialize_params would be more suitable

I agree that `reset_params` is not the best name. My opinion is that `reinitialize_params` is the best alternative. Even though `reinitializable` sounds a bit awkward, its no worse than `regularizable`. due to the -able suffix. My only other idea is reinitable, but that isn't consistent with other standard tags.  

> My main objection is that adding a resetting feature that only resets some of the weights may lead users into running wrong experiments. 

I agree that there may be a bit of confusion here, but I think the tradeoff is worth it. I also think that using `reinitialize_params` as the function name will help mitigate this confusion. Between this and good documentation, I think the downside is low. 

On lazy initialization:
> If implemented, this should be optional.

Not only should it be optional it should be off by default. However, given that params are allowed to be reinitializable, then it seems like it would be a natural extension to allow them to be lazy-initializable. This increases my belief that this feature is a move in the right direction.  

I'll also continue in #740

**Comment by astooke** [comment]
This all sounds interesting...being able to reinitialize parameter values in layers that were initialized using a Lasagne initializer would be helpful :)

In the meantime, I will instantiate a new network, get the values out of that, and set the original network's values to those, to keep all the same Theano variables in play.  Luckily my networks are small, so it's not a problem to have the new network instantiated on the GPU.  I supposed if that were an issue, I would have to make a bunch of copies of the network on CPU, keep all their values around as the future reinitializers, and then make the `theano.gpuarray.use()` call.
